### ISSUE TITLE
2016.3.1 Salt-API Error

### ISSUE THREAD

**Issue by deuscapturus** [saltstack/salt#34234]
### Description of Issue

Since upgrading to salt 2016.3.1 the salt-api gets into a failing state on a daily basis.  Restarting salt-api resolves the issue for a while.

Error in log:

```
2016-06-23 04:24:28,108 [cherrypy.error.55417232                              ][ERROR   ][11147] [23/Jun/2016:04:24:28] HTTP Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/cherrypy/_cprequest.py", line 666, in respond
    self.hooks.run('before_finalize')
  File "/usr/lib/python2.6/site-packages/cherrypy/_cprequest.py", line 112, in run
    raise exc
KeyError: u'f4e403408e9a9b1821633c02bd2818cef8859bfe'

2016-06-23 04:24:28,109 [cherrypy.error.55417232                              ][ERROR   ][11147] [23/Jun/2016:04:24:28]  Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/cherrypy/_cprequest.py", line 102, in run
    hook()
  File "/usr/lib/python2.6/site-packages/cherrypy/_cprequest.py", line 62, in __call__
    return self.callback(**self.kwargs)
  File "/usr/lib/python2.6/site-packages/cherrypy/lib/sessions.py", line 706, in close
    sess.release_lock()
  File "/usr/lib/python2.6/site-packages/cherrypy/lib/sessions.py", line 374, in release_lock
    self.locks[self.id].release()
KeyError: u'f4e403408e9a9b1821633c02bd2818cef8859bfe'
```

The client hitting the salt-api gets this reponse:

```
Error Getting Token: <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"></meta>
    <title>500 Internal Server Error</title>
    <style type="text/css">
    #powered_by {
        margin-top: 20px;
        border-top: 2px solid black;
        font-style: italic;
    }

    #traceback {
        color: red;
    }
    </style>
</head>
    <body>
        <h2>500 Internal Server Error</h2>
        <p>The server encountered an unexpected condition which prevented it from fulfilling the request.</p>
        <pre id="traceback"></pre>
    <div id="powered_by">
    <span>Powered by <a href="http://www.cherrypy.org">CherryPy 3.2.2</a></span>
    </div>
    </body>
</html>
```
### Steps to Reproduce Issue

(Include debug logs if possible and relevant.)
### Versions Report

```
Salt Version:
           Salt: 2016.3.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.2.2
       dateutil: 2.1
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: 0.20.0
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: 0.20.3
         Python: 2.6.9 (unknown, Dec 17 2015, 01:08:55)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist:   
        machine: x86_64
        release: 4.1.13-19.30.amzn1.x86_64
         system: Linux
        version: Not Installed
```

**Comment by DrMerlin** [comment]
@markahopper

**Comment by Ch3LL** [comment]
@deuscapturus  or @DrMerlin  if you upgrade cherrypy to 3.6.0 are you still seeing this behavior? I noticed [this](https://bitbucket.org/cherrypy/cherrypy/pull-requests/50/fix-race-condition-in-session-clean-up/diff) fix for the same error for cherrypy that looks like it was made availabe in 3.6.0 in the changelog [here](https://github.com/cherrypy/cherrypy/blob/master/CHANGES.txt#L146)

**Comment by deuscapturus** [comment]
We believe this issue is actually related to this: https://github.com/saltstack/salt/issues/34215

**Comment by Ch3LL** [comment]
I cannot replicate this error as well using the master config posted in that issue you think its related to. Where you able to upgrade your cherrypy version and see if that helped?

**Comment by Ch3LL** [comment]
also ping @whiteinge do you have any ideas why this might be happening or any ideas on how to replicate this behavior.?

**Comment by whiteinge** [comment]
Hm. The `KeyError` comes from trying to look up the session token (in order to grab the Salt eauth token). I haven't seen this error before and 3.2.2 is well-tread territory. Few questions:

* How long do you reuse the same session token? By default it only lasts the same length as Salt's eauth token which is the `token_expire` setting in the master config. After that you need a new token.
* salt-api will lose any session tokens whenever it is restarted. Is it getting restarted during the day either on purpose or by accident?

**Comment by Ch3LL** [comment]
@deuscapturus also we would like to know why you believe these two issues are related? 

By chance are you possibly seeing the salt-api and master crash at the same time? or any other indications?

**Comment by whiteinge** [comment]
Echoing the detail above this appears to be a race condition in the CherryPy server that was fixed for the 3.2.5 release (at quick glance, according to the Git history). Unfortunately there was an SSL-related problem for several CherryPy releases between 3.2.5 and 3.7.x and possibly more. The current version 10.1.x appears to be fully working so I'd feel comfortable closing this out once we have a few more testimonials for that version range.

**Comment by dmurphy18** [comment]
@whiteinge @Ch3LL the 10.2.1 version supports Python 2.7 through 3.5.  Did they install this on Redhat 6, which is python 2.6 ?, which implies issues regardless of if they got it working, and use satisfactorily packing it.
